New module for managing NetApp E-Series iSCSI Interfaces

lmprice opened the pull request:
Define a new module for configuring NetApp E-Series iSCSI interfaces.

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

Introduce a new module for managing E-Series iSCSI interfaces. Each E-Series system that supports iSCSI will have between 2 and 6 iSCSI interfaces (channels), that can be configured on each controller.

##### ISSUE TYPE
 - New Module Pull Request
##### COMPONENT NAME
netapp_e_iscsi_interface

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.6.0 (detached HEAD ce8c336d52) last updated 2018/04/18 11:23:03 (GMT -500)
  config file = None
  configured module search path = [u'/home/lorenp/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/lorenp/ansible/lib/ansible
  executable location = /home/lorenp/ansible/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
changed: [testhost] => (item=1) => {"changed": true, "enabled": true, "item": 1, "msg": "The interface settings have been updated."}
```

lmprice committed:
New module for NTAP E-Series iSCSI Interfaces

Define a new module for configuring NetApp E-Series iSCSI interfaces.

mattclay commented on test/integration/integration_config.yml:
```diff
@@ -1,3 +1,7 @@
+netapp_e_ssid: 1
+netapp_e_api_host: 10.113.1.111:8443
+netapp_e_api_username: admin
+netapp_e_api_password: infiniti
```
Integration tests should not add anything to the shared integration config.

lmprice commented:
@mattclay That issue should be resolved. Not sure how I can make the request resolved.

hulquest commented on lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:
```diff
@@ -0,0 +1,379 @@
+    def interfaces(self):
+        ifaces = list()
+        try:
+            (rc, ifaces) = request(self.url + 'storage-systems/%s/graph/xpath-filter?query=/controller/hostInterfaces'
```
This looks like a newer endpoint.  Does the documentation need to have a minimum version of SANtricity OS?

lmprice replied:
@hulquest We added this back in the WSP 1.3 time-frame (early 2016). While it's probably been in there long enough it doesn't need to be documented (IMO), that does beg the question of how to do so for when it does. I can add a note to the module if that makes sense.

hulquest replied:
I think a brief sentence is worthy.  Nothing is worse than trying something only to figure out the target API on the server isn't compatible.  We should try to remember to ask this question of all new modules and substantial changes.

lmprice replied:
@hulquest I added a note in the documentation. In the next patch I submit I'll see about adding a more lengthy blurb in the doc_fragments about API requirements, etc.

gundalow replied:
@hulquest @lmprice Is this discussion resolved?

lmprice replied:
@gundalow Yes, I believe that it has. I added a comment in the notes about what API version is required for this module.

hulquest reviewed (commented):
Once the question about backward compatibility is answered, this looks good to me.

Commit:
Improve netapp_e_iscsi_interface integration tests

Restructured integration test to set all iscsi ports to disabled, then
defines the ports either statically or with dhcp, next updates the ports
with the other definition type (static <-> dhcp), and lastly disables
all ports.  Each netapp_eseries_iscsi_interface call is verified with the
array.

lmprice commented:
@hulquest @amit0701 @schmots1 @carchi8py Can you give me a review on this patch?

hulquest commented:
shipit

lmprice commented:
@hulquest Just a theory, but can you hit review and mark it as "approved"?

ndswartz reviewed (approved):
shipit

gundalow commented on lib/ansible/utils/module_docs_fragments/netapp.py:
```diff
@@ -136,7 +136,8 @@ class ModuleDocFragment(object):
     required: false
```
Please remove this file from the PR

gundalow commented on test/sanity/validate-modules/ignore.txt:
```diff
@@ -1133,7 +1133,6 @@ lib/ansible/modules/storage/netapp/na_cdot_volume.py E322
 lib/ansible/modules/storage/netapp/na_cdot_volume.py E325
 lib/ansible/modules/storage/netapp/netapp_e_amg.py E322
-lib/ansible/modules/storage/netapp/netapp_e_amg.py E325
 lib/ansible/modules/storage/netapp/netapp_e_amg_role.py E322
```
Please remove this file from the PR

gundalow commented on lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:
```diff
@@ -0,0 +1,380 @@
+        description:
+            - The IPv4 address to assign to the interface.
+            - Should be specified in xx.xx.xx.xx form.
+            - Mutually exclusive with I(config_method=dhcp)
```
Add code to check for this and do `fail_json` if needed just after `argument_spec`

gundalow commented on lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:
```diff
@@ -0,0 +1,380 @@
+        description:
+            - The subnet mask to utilize for the interface.
+            - Should be specified in xx.xx.xx.xx form.
+            - Mutually exclusive with I(config_method=dhcp)
```
Add code to check for this and do `fail_json` if needed just after `argument_spec`

gundalow commented on lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:
```diff
@@ -0,0 +1,380 @@
+        description:
+            - The IPv4 gateway address to utilize for the interface.
+            - Should be specified in xx.xx.xx.xx form.
+            - Mutually exclusive with I(config_method=dhcp)
```
Add code to check for this and do `fail_json` if needed just after `argument_spec`

gundalow commented on lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:
```diff
@@ -0,0 +1,380 @@
+    config_method:
+        description:
+            - The configuration method type to use for this interface.
+            - dhcp is mutually exclusive with I(address), I(subnet_mask), and I(gateway).
```
Add code to check for this and do `fail_json` if needed just after `argument_spec`

lmprice commented:
@gundalow The issues should be resolved.
